librdmacm/rsocket.c

```diff
@@ -1453,9 +1453,14 @@ static int rs_do_connect(struct rsocket *rs)
 
 		rs->state = rs_connect_rdwr;
 		break;
+	case rs_connect_error:
+	case rs_disconnected:
+	case rs_error:
+		ret = ERR(ENOTCONN);
+		goto unlock;
 	default:
 		ret = (rs->state & rs_connected) ? 0 : ERR(EINVAL);
-		break;
+		goto unlock;
```

Comment:
Is it OK to not change the state to rs_connect_error on both of these unwinds?

Reply:
I believe so.  In the added code, the socket is in the wrong state for processing, so the call just returns an error.    In the other cases, we've handled an event that's driving the socket state.  If in the processing of that event an error occurs, the connection failed (connect_error).

If we are in the disconnected or error state, then some other thread finished establishing the connection and tore it down before this thread could execute.  So it would be incorrect to change the state to report a connect_error.

I think the only viable state we can be in in the default case now is connected.